### PR TITLE
Add webhooks v3

### PR DESCRIPTION
**Pull request by melnychukJS**

### CODE REVIEW

File: source/includes/reference/_app_webhooks.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
+# Application Webhooks
+
+Application webhook subscription is a mechanism that will allow you to receive notifications to your servers whenever various events happen in relation to different resources created by an ***application***. You can subscribe to receive webhooks about resource created under a specific [***profile***](#profile-webhooks).
+
+Before proceeding, make sure the endpoint where you intend to receive webhooks satisfies the following requirements:
+
+* Has a valid domain name (IPs are disallowed)
```
Comment (henry-fraser) [inline source/includes/reference/_app_webhooks.md]: Minor, but: `IP addresses`

File: source/includes/reference/_app_webhooks.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+* Listens to HTTPS requests on port `443`
+* Has a valid HTTPS certificate signed by a trusted Certificate Authority - CA (self-signed or expired certificates are not accepted)
+* Does not include any query arguments in the URL
```
Comment (henry-fraser) [inline source/includes/reference/_app_webhooks.md]: Do we actually check that?

File: source/includes/reference/_app_webhooks.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+`https://webhooks.example.com/balance-change` is a valid URL; `http://webhooks.example.com:8080/hook.php?type=balance` is not.
+
+You can have multiple subscriptions per event type though be mindful you will receive duplicate callbacks, one for each subscription.
+
+> **NB! Please note that you have to use client level token for making following requests.**
```
Comment (henry-fraser) [inline source/includes/reference/_app_webhooks.md]: Pedantic mode: "NB" basically means "please note"

Comment (henry-fraser) [reply source/includes/reference/_app_webhooks.md]: "client level tokens"

File: source/includes/reference/_app_webhooks.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    "trigger_on": "transfers#state-change", 
+    "scope": {
+        "domain": "application", 
+        "id": "<your client key>"
+    },
+    "created_by": {
+        "type": "application",
+        "id": "<your client id>", //clientId and key are not always the same
+    },
+    "created_at": "2008-09-15T15:53:00+00:00Z",
```
Comment (henry-fraser) [inline source/includes/reference/_app_webhooks.md]: Not sure it matters, but that timestamp does not look correct.  Either `+00:00` or `Z`, pick one?

File: source/includes/reference/_app_webhooks.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+**`POST https://api.transferwise.com/v3/applications/{clientKey}/subscriptions`**
+
+`{clientKey}` can be received upon obtaining client credentials from our tech support.
+
+All fields listed below are required for creating a webhook subscription.
+
+Field                     | Description                                                             | Format
+---------                 | -------                                                                 | -----------
+name                      | A custom name for your webhook to ease with identification              | Text
+trigger_on                | `transfers#state-change`, `transfers#active-cases` or `balances#credit` | Text
```
Comment (henry-fraser) [inline source/includes/reference/_app_webhooks.md]: This list is likely to change: let's link from here to another table that lists them all and can be easily extended.

File: source/includes/reference/_app_webhooks.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+delivery.version          | The event representation semantic                                       | Text
```
Comment (henry-fraser) [inline source/includes/reference/_app_webhooks.md]: "semantic version".

If we have a separate table for the event types, we can detail the various versions.

File: source/includes/reference/_app_webhooks.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+
+### Response
+
+Field                     | Description                                                             | Format
+---------                 | -------                                                                 | -----------
+id                        | UUID that uniquely identifies the subscription                          | Text
+name                      | A custom name for your webhook to ease with identification              | Text
+trigger_on                | `transfers#state-change`, `transfers#active-cases` or `balances#credit` | Text
+delivery.version          | The event representation semantic                                       | Text
```
Comment (henry-fraser) [inline source/includes/reference/_app_webhooks.md]: As above...

File: source/includes/reference/_app_webhooks.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+delivery.url              | Required. The URL where your server will be listening for events.       | Text
+scope.domain              | Scope of this subscription, always "application" in this case           | Text
+scope.id                  | Client key used to create this subscription                             | Text
+created\_by.type          | Creator type. Always application in this case                           | Text
+created\_by.id            | Client id of the creator. Not always the same as the client key         | Text
```
Comment (henry-fraser) [inline source/includes/reference/_app_webhooks.md]: Very minor, but I would always capitalise ID in phrases such as "client ID".

File: source/includes/reference/_profile_webhooks.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+**`POST https://api.transferwise.com/v3/profiles/{profileId}/subscriptions`**
+
+`profileId` - ID of the profile you are subscribing to. 
+
+All fields listed below are required for creating a webhook subscription.
+
+Field                     | Description                                                             | Format
+---------                 | -------                                                                 | -----------
+name                      | A custom name for your webhook to ease with identification              | Text
+trigger_on                | `transfers#state-change`, `transfers#active-cases` or `balances#credit` | Text
```
Comment (henry-fraser) [inline source/includes/reference/_profile_webhooks.md]: As mentioned, let's link to some other shared section that

- describes the events
- details each version and their structure

Not sure that exists already.

File: source/includes/reference/_webhooks.md [hunk 10]
```diff
@@ -1,89 +1,14 @@
 # Webhooks
 
+**NB! Please note this version of webhooks is now **_deprecated_**. You can use (profile webhooks) or (application webhoosk) instead.
```
Comment (henry-fraser) [inline source/includes/reference/_webhooks.md]: webhooks

File: source/includes/reference/_app_webhooks.md (hunk 3 above)
Comment (mce-transferwise) [reply source/includes/reference/_app_webhooks.md]: is it worth explaining what a client level token is? in the docs it's only mentioned in one place: https://api-docs.transferwise.com/#users-sign-up-with-registration-code

File: source/includes/reference/_app_webhooks.md (hunk 5 above)
Comment (mce-transferwise) [reply source/includes/reference/_app_webhooks.md]: also `balances#credit`  is not available for applications. Technically they can subscribe but they will not receive any webhook.

File: source/includes/reference/_app_webhooks.md [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+### Response
+
+Field                     | Description                                                             | Format
+---------                 | -------                                                                 | -----------
+id                        | UUID that uniquely identifies the subscription                          | Text
+name                      | A custom name for your webhook to ease with identification              | Text
+trigger_on                | `transfers#state-change`, `transfers#active-cases` or `balances#credit` | Text
+delivery.version          | The event representation semantic                                       | Text
+delivery.url              | Required. The URL where your server will be listening for events.       | Text
```
Comment (mce-transferwise) [inline source/includes/reference/_app_webhooks.md]: minor: as this is response 'Required' can be removed.

File: source/includes/reference/_app_webhooks.md [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+> Example Request:
+
+```shell
+curl -X DELETE "https://api.transferwise.com/v3/applications/{clientKey}/subscriptions/{id}" \
+  -H "Authorization: Bearer <your client level token>"
+```
+
+> Example Response:
+
+```json
```
Comment (mce-transferwise) [inline source/includes/reference/_app_webhooks.md]: i thought we respond with 204 (NO_CONTENT) without a body

File: source/includes/reference/_app_webhooks.md (hunk 2 above)
Comment (melnychukJS) [reply source/includes/reference/_app_webhooks.md]: Yes, https://github.com/transferwise/webhook-sender/blob/master/src/main/java/com/transferwise/webhook/services/callback/StrictUrlValidationService.java#L29

File: source/includes/reference/_app_webhooks.md (hunk 12 above)
Comment (melnychukJS) [reply source/includes/reference/_app_webhooks.md]: Will test this

File: source/includes/reference/_app_webhooks.md (hunk 5 above)
Comment (melnychukJS) [reply source/includes/reference/_app_webhooks.md]: Will include this note on the events page, thanks for flagging!

File: source/includes/reference/_profile_webhooks.md [hunk 13]
```diff
@@ -0,0 +1,204 @@
+# Profile Webhooks
+
+Profile webhook subscription is a mechanism that will allow you to receive notifications to your servers whenever various events happen in relation to different resources created under a specific ***profile***. You can subscribe to receive webhooks about resource created under a specific [***application***](#app-webhooks).
```
Comment (mce-transferwise) [inline source/includes/reference/_profile_webhooks.md]: > You can subscribe to receive webhooks about resource created under a specific [***application***](#app-webhooks)

Should `specific [***application***]` be `specific [***profile***]` ? If yes i think we need to update similar statement above

Comment (melnychukJS) [reply source/includes/reference/_profile_webhooks.md]: Hmm, my goal was to indicate that this is the docs for profile subscriptions, but there is also an option for creating application subscriptions (hence the link)

File: source/includes/reference/_profile_webhooks.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+
+```shell
+curl -X DELETE "https://api.transferwise.com/v3/profiles/{profileId}/subscriptions/{id}" \
+  -H "Authorization: Bearer <your user token>"
+```
+
+> Example Response:
+
+```json
+{
```
Comment (andriimelnychuk) [inline source/includes/reference/_profile_webhooks.md]: So what was the outcome of the test do we return 204 with empty body?

Comment (melnychukJS) [reply source/includes/reference/_profile_webhooks.md]: Yes, we return 204 with empty body

**Pull request merged**